### ISSUE TITLE
Facade coverage: three model shapes cannot be expressed through AiModelBuilder/AiModelResult

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#2101]
## Summary

Converting the documentation examples to the facade proved which model shapes it cannot express. This is
the concrete answer to "can a user do everything through `AiModelBuilder` and `AiModelResult`?"

**For ordinary supervised train-and-predict: yes.** 137 examples were converted and compile, covering
regression, classification, boosting, ensembles and neural networks.

**For three shapes: no.** Each was attempted, failed to compile, and was reverted to direct model use.

## 1. Models that train on three inputs (13 examples)

`Build` takes `(features, labels)`. These take three, and there is no overload for them:

| area | signature |
|---|---|
| Causal inference — `CausalForest`, `SLearner`, `TLearner`, `XLearner`, `DoublyRobustEstimator`, `InverseProbabilityWeighting`, `PropensityScoreMatching` | `Train(features, treatment, outcome)` |
| Survival analysis — `CoxProportionalHazards`, `WeibullAFT`, `LogNormalAFT`, `RandomSurvivalForest` | `Train(times, events, features)` |
| `MedGANGenerator` | `Fit(data, columns, epochs)` |

Treatment assignment and censoring indicators are not labels, so folding them into the existing two
parameters would misrepresent them.

## 2. Models with an API beyond predict (8 examples)

`AiModelResult.Model` is **internal**:

```csharp
internal IFullModel<T, TInput, TOutput>? Model { get; set; }
```

So once a model goes through the builder, anything it offers beyond `Predict` becomes unreachable:

- `STLDecomposition.GetTrend()`, `.GetSeasonal()`, `.GetResidual()` — the decomposition IS the output
- `ARIMAModel.Forecast(horizon)`, `DynamicRegressionWithARIMAErrors`
- `BayesianStructuralTimeSeriesModel`, `UnobservedComponentsModel`, `SpectralAnalysisModel`
- `KaplanMeierEstimator`, `NelsonAalenEstimator` — survival curves

For a decomposition model, `Predict` is not what the user came for. Routing it through the facade would
delete the reason the type exists.

## 3. Unsupervised models (1 example, but a whole category)

`SpectralClustering` has no labels. `Build(features, labels)` requires them, so clustering cannot be built
through the facade at all. The same applies to every clustering and dimensionality-reduction type; only
one surfaced here because the others' examples fail for unrelated reasons and were not yet converted.

## Suggested scope

- A `Build` overload for unsupervised fitting — `Build(features)` — which also covers dimensionality
  reduction.
- A way to reach model-specific output through the result. Making `Model` public is the blunt version and
  gives up the encapsulation; a typed accessor such as `result.As<STLDecomposition<double>>()`, or
  surfacing decomposition and forecast results on `AiModelResult` directly, keeps the facade as the single
  entry point.
- Either a three-input `Build` overload, or first-class treatment/censoring parameters, so causal and
  survival models are reachable.

## Enforcement already in place

`docs-wiki.yml` now runs a facade-conformance gate: `DocSnippetVerify --facade-check` resolves the
receiver of every `Train`/`Predict`/`Fit` call through Roslyn's semantic model and fails when the count of
non-facade calls rises. The ceiling is 318 and ratchets down. Constructing a model is deliberately not a
violation — `ConfigureModel` takes an `IFullModel` — only calling through it is.

**Comment by ooples** [comment]
**Part 2 is done in #2088; parts 1 and 3 are not.** Leaving this open for the remaining two.

## Part 2 — model-specific output — resolved

Took the option this issue preferred over making `Model` public: 13 accessors on `AiModelResult`, each
type-testing the wrapped model and delegating, and throwing `NotSupportedException` naming what was
actually built when the model is the wrong kind.

| partial | accessors |
|---|---|
| `AiModelResult.CausalEffects` | `EstimateATE`, `EstimateATT`, `EstimateTreatmentEffect` |
| `AiModelResult.TimeSeriesComponents` | `GetTrend`, `GetSeasonal`, `GetFrequencies`, `GetPeriodogram`, `ComputeAnomalyScores`, `Forecast` ×2 |
| `AiModelResult.ModelSpecific` | `GetClusterLabels`, `AdaptAndEvaluate`, `GenerateSamples`, `DeriveArchitecture` |

`FourierNeuralOperator.Forward` needed no accessor — it derives from `NeuralNetworkBase`, so
`result.Predict` already covered it.

Covered by `AiModelResultAccessorTests` (27 tests, two per accessor: delegation and guard), verified
non-vacuous by mutating `GetTrend` to delegate to `GetSeasonal`.

## Part 1 — three-input training — examples compile, the API concern stands

The examples were converted, but by using conventions that already existed rather than by adding an
overload:

- **Causal.** `CausalModelBase.Train(x, y)` already documents that column 0 of `x` is the binary
  treatment indicator and columns 1.. are the covariates, and dispatches to the three-argument `Fit`.
  So the builder path works today.
- **Survival.** `SurvivalModelBase.Train(x, y)` reads `y` as the observed times and sets every event
  indicator to 1 — it silently assumes no censoring, which is the whole point of survival analysis.
- **MedGAN.** Routed through the tensor `Build` instead of the column-metadata `Fit`.

This issue's objection — that treatment and censoring are not labels and folding them in misrepresents
them — is unchanged. The survival case is the sharpest: the facade path quietly discards censoring.

## Part 3 — unsupervised models — worked around, not fixed

`Build(features)` still does not exist. The clustering examples pass an ignored label vector and say so:

```csharp
// clustering is unsupervised, so the builder's label argument is ignored
var unused = new Vector<double>(dataMatrix.Rows);
```

That compiles and is honest, but it is a workaround. `SpectralClustering`, `KMeans`, `BIRCH`, `CURE`,
`CLARANS`, `GaussianMixtureModel` and `SelfOrganizingMap` all read this way now.

## Enforcement

The ceiling this issue mentions at 318 is now **0**: `--facade-check --max-facade-violations 0`, with
1415/1415 XML and 1700/1700 wiki examples compiling. Any new non-facade call fails CI.

**Comment by ooples** [comment]
**Part 1 is now half done: survival has a three-argument `Build`; causal still uses the column-0 convention.**

## Survival — `Build(features, times, events)` (#2088)

```csharp
var result = new AiModelBuilder<double, Matrix<double>, Vector<double>>()
    .ConfigureModel(new KaplanMeierEstimator<double>())
    .Build(features, times, events);      // events: 1 = observed, 0 = censored
```

Declared on `IAiModelBuilder<T, TInput, TOutput>` so it survives a `ConfigureModel` chain, and guarded
against a non-survival model, a missing model, mismatched lengths, a non-binary indicator, and a builder
over the wrong input/output types — each naming what was actually configured.

Two things worth recording, because the obvious implementation is wrong:

**It does not pack the indicator into the covariate matrix.** That was the first attempt. It trains
fine and then fails on the first `Predict`: the result remembers a column the caller cannot supply,
because whether the event occurred is exactly what prediction is for. The indicators are handed to the
model alongside the covariates instead, consumed by the next `Train` and cleared.

**Rows must not be shuffled or split.** The indicators are one value per row, so an 80% shuffled subset
pairs each subject with somebody else's censoring. Survival-with-events therefore joins clustering on
the existing `useFullData` path — same reason, different shape — and models built this way train on
every row.

Also fixed underneath it: `SurvivalModelBase.Train(x, y)` used to set every event indicator to 1,
silently asserting nobody was censored (#2115).

## Causal — still the column-0 convention

`CausalModelBase.Train(x, y)` reads column 0 as the binary treatment indicator, and the seven causal doc
examples assemble that design matrix by hand. It works and is documented, but carries the same residual
risk the survival path just shed: a covariate left in column 0 is read as treatment.

The survival machinery generalises — a `SupplyTreatment` alongside `SupplyEvents`, the same
`useFullData` reasoning, and a `Build(covariates, treatment, outcome)` overload. Worth noting that the
two overloads would be distinguished only by which argument is `Vector<int>`:

```csharp
Build(Matrix<T> features,   Vector<T> times,     Vector<int> events)     // survival
Build(Matrix<T> covariates, Vector<int> treatment, Vector<T> outcome)    // causal
```

That is type-safe for any realistic `T` — you cannot call one meaning the other — but the two collide
when `T` is `int`. Distinct names would avoid it at the cost of an asymmetry with the survival overload.

## Part 3 — unsupervised `Build(features)` — still open

Unchanged: clustering examples pass an ignored label vector and say so.

**Comment by ooples** [comment]
**Part 1 is done.** Both domains now have a three-argument `Build`; only part 3 (unsupervised) remains.

```csharp
.Build(features, times, events)          // survival: 1 = observed, 0 = censored
.Build(covariates, treatment, outcome)   // causal:   1 = treated,  0 = control
```

Both are declared on `IAiModelBuilder<T, TInput, TOutput>` so they survive a `ConfigureModel` chain, and
both hand the third input to the model rather than folding it into the covariate matrix — which was the
first implementation and is wrong in both domains, because the covariate matrix has to stay itself on
the way *out* too. Survival prediction takes covariates and the event is what is being predicted; the
causal effect estimators take covariates and treatment separately.

The two are told apart by which argument is the `Vector<int>` indicator, so passing them the wrong way
round is a compile error rather than a silently different call. When `T` is `int` they collide — also a
compile error, at the call site, verified with a probe rather than assumed, and neither domain has a use
for integer times or outcomes.

Models handed a third input take the direct training path on the full, unshuffled data. That matters:
the indicator is one value per row, so the optimizer's repeated fits over candidate subsets would pair
each subject with somebody else's treatment or censoring. Same clause clustering already had.

Two defects surfaced underneath the work, both fixed and filed:

- **#2115** — `SurvivalModelBase.Train(x, y)` set every event indicator to 1, silently asserting nobody
  was censored.
- **#2116** — `SLearner`, `XLearner`, `InverseProbabilityWeighting` and `DoublyRobustEstimator` could
  not be built through the facade at all, on the documented two-argument path as much as the new one.
  `SupportsParameterInitialization` answered with `ParameterCount`, which throws on a fit-deferred
  layout, and the builder probes it to choose a training path.

## Still open: part 3 — unsupervised `Build(features)`

Unchanged. Clustering examples pass an ignored label vector and say so:

```csharp
// clustering is unsupervised, so the builder's label argument is ignored
var unused = new Vector<double>(dataMatrix.Rows);
```

`SpectralClustering`, `KMeans`, `BIRCH`, `CURE`, `CLARANS`, `GaussianMixtureModel` and
`SelfOrganizingMap` all read this way. Clustering already takes the full-data path, so a
`Build(features)` overload would be the smallest of the three.

**Comment by ooples** [comment]
**All three parts are done.** Every model shape this issue named is now expressible through
`AiModelBuilder` / `AiModelResult`.

| part | shape | resolution |
|---|---|---|
| 1 | three-input training | `Build(features, times, events)` and `Build(covariates, treatment, outcome)` |
| 2 | API beyond predict | 13 accessors on `AiModelResult` |
| 3 | unsupervised models | `Build(features)` |

## Part 3 — `Build(features)`

```csharp
var result = new AiModelBuilder<double, Matrix<double>, Vector<double>>()
    .ConfigureModel(new KMeans<double>(new KMeansOptions<double> { NumClusters = 2 }))
    .Build(dataMatrix);

var assignments = result.GetClusterLabels();
```

The label argument was never used by these models, so callers constructed one and explained it away:

```csharp
// clustering is unsupervised, so the builder's label argument is ignored
var unused = new Vector<double>(dataMatrix.Rows);
```

A parameter that exists only to be discarded is worth removing rather than documenting. The overload
builds the placeholder itself, shaped to the builder's output type so a tensor-typed builder is served
as readily as a vector-typed one, and refuses a supervised model rather than quietly discarding labels
it would have used.

All seven clustering examples — `KMeans`, `BIRCH`, `CURE`, `CLARANS`, `GaussianMixtureModel`,
`SpectralClustering`, `SelfOrganizingMap` — now read as `Build(dataMatrix)`.

## The enforcement this issue proposed

The facade-conformance gate it describes with a ceiling of 318 now runs at **0**, alongside 1418/1418
XML and 1700/1700 wiki examples compiling. Any new non-facade call fails CI.

## Related defects found along the way

Three of the four were invisible until a caller tried to use the shape this issue asked for:

- **#2115** — survival models silently discarded censoring.
- **#2116** — four causal models could not be built through the facade at all.
- **#2113** — causal standard errors were not reproducible.
- **#2114** — a dead public `Model` property on `AiModelResultOptions`.

Closing. Part 3's `Build(features)` covers clustering; if dimensionality-reduction models are added to
the builder later they will want the same overload, which they will get for free by deriving from a base
the guard recognises.
